docs/hackathons/ihp_10_2022.rst

```diff
@@ -0,0 +1,285 @@
+
+
+* **combine different numerical algorithms**. After having different algorithms implemented as small objects, we can start combining them using composition. For example, we can have a less accurate (but fast) ``LogSolver`` to create the initialization for a more accurate ``LogSolver``. Hard coding them in a ``Connection`` would have never allowed this kind of combinations.
+
+
+
+For a contributor, setting a (default) numerical algorithm follows a similar strategy as ``_define_default_metric`` shown above. The only difference is that it should return an instance, instead of the class. For example:
+
+
+.. code-block:: python
+
+    class MyMetric(RiemannianMetric):
+
+        def _define_exp_solver(self):
+         return MyNiceExpSolver(n_steps=10)
+
+
+If ``_define_exp_solver`` does not exist, we will assume a closed-form solution is implemented. A nice error message will be implemented in ``Connection.exp`` to help contributors know what to change if something goes wrong.
+
+
+
+gs.matrices
+***********
+
+``Matrices``, ``SymmetricMatrices``, and ``SPDMatrices`` have several linear algebra-related (class) methods that are useful in several parts of the code. Having to import them any time we need to perform these operations increases the likelihood of circular imports and makes everything messier. Besides, these methods are a bit hidden within these classes as ``geomstats.geometry`` is supposed to handle geometric operations. We will then move them to the backend (under ``gs.matrices``), ensure proper testing, and add backend documentation.
+
+
+One question that arises with this is that some methods are particular to specific kinds of matrices. We will follow ``numpy``/``scipy`` conventions to handle this (e.g. ``eig`` vs ``eigh``). Moreover, we will have a ``check`` boolean parameter set to ``False`` by default, meaning we will not check if a user that is calling ``eigh`` is really passing a hermitian matrix, as these checks are computationally expensive (the possibility of ``check=True`` is especially attractive during development/debugging).
+
+
+
+
+Points shape
+************
+
+Methods in ``geomstats.geometry`` will still be able to receive one point or multiple points, but some behaviors will be slightly modified. In particular, the following behaviors are expected for a method that receives a 1d point and returns a 1d point (after some operations):
+
+1. input: 1d -> output: 1d
+
+2. input: 2d (multiple points) -> output: 2d (multiple points)
+
+3. input: 2d (one point) -> output: 2d (one point)
+
+
+The last behavior is new, as currently if we have one point as input, we will always return a 1d array (for the 1d-1d case). This means ``gs.squeeze`` will have to be used at a much lower extent.
```

Comment:
Good point, although I would argue that the current behavior is not consistent across the code base.

Reply:
Unfortunately it is not. The idea is to set a rule and now ensure it is followed everywhere.